Hum, 2 tests failed on the macOS :-( @corona10: Can you have a look?

```
======================================================================
ERROR: test_imaplib_timeout_test (test.test_imaplib.NewIMAPSSLTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/test/test_imaplib.py", line 446, in test_imaplib_timeout_test
    client = self.imap_class("localhost", addr, timeout=support.LOOPBACK_TIMEOUT)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 1307, in __init__
    IMAP4.__init__(self, host, port, timeout)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 201, in __init__
    self.open(host, port, timeout)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 1320, in open
    IMAP4.open(self, host, port, timeout)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 311, in open
    self.sock = self._create_socket(timeout)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 1311, in _create_socket
    return self.ssl_context.wrap_socket(sock,
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/ssl.py", line 500, in wrap_socket
    return self.sslsocket_class._create(
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/ssl.py", line 1040, in _create
    self.do_handshake()
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/ssl.py", line 1309, in do_handshake
    self._sslobj.do_handshake()
socket.timeout: _ssl.c:1091: The handshake operation timed out


======================================================================
ERROR: test_imaplib_timeout_test (test.test_imaplib.NewIMAPTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/test/test_imaplib.py", line 446, in test_imaplib_timeout_test
    client = self.imap_class("localhost", addr, timeout=support.LOOPBACK_TIMEOUT)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 204, in __init__
    self._connect()
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 246, in _connect
    self.welcome = self._get_response()
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 1058, in _get_response
    resp = self._get_line()
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 1166, in _get_line
    line = self.readline()
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/imaplib.py", line 322, in readline
    line = self.file.readline(_MAXLINE + 1)
  File "/Users/runner/runners/2.163.1/work/cpython/cpython/Lib/socket.py", line 704, in readinto
    return self._sock.recv_into(b)
socket.timeout: timed out
```

--- a/Doc/library/imaplib.rst
+++ b/Doc/library/imaplib.rst
@@ -369,7 +369,9 @@ An :class:`IMAP4` instance has the following methods:
    Opens socket to *port* at *host*. The optional *timeout* parameter
    specifies a timeout in seconds for the connection attempt.
    If timeout is not given or is None, the global default socket timeout
-   is used. This method is implicitly called by the :class:`IMAP4` constructor.
+   is used. Also note that if the *timeout* parameter is set to be zero,
+   it will raise a :class:`ValueError` to reject creating a non-blocking socket.
+   This method is implicitly called by the :class:`IMAP4` constructor.
    The connection objects established by this method will be used in
    the :meth:`IMAP4.read`, :meth:`IMAP4.readline`, :meth:`IMAP4.send`,
    and :meth:`IMAP4.shutdown` methods. You may override this method.

--- a/Doc/library/imaplib.rst
+++ b/Doc/library/imaplib.rst
@@ -30,7 +30,7 @@ Three classes are provided by the :mod:`imaplib` module, :class:`IMAP4` is the
 base class:
 
 
-.. class:: IMAP4(host='', port=IMAP4_PORT[, timeout=None])
+.. class:: IMAP4(host='', port=IMAP4_PORT, timeout=None)
 
    This class implements the actual IMAP4 protocol.  The connection is created and
    protocol version (IMAP4 or IMAP4rev1) is determined when the instance is
@@ -366,14 +366,13 @@ An :class:`IMAP4` instance has the following methods:
 
 .. method:: IMAP4.open(host, port, timeout=None)
 
-   Opens socket to *port* at *host*.  This method is implicitly called by
-   the :class:`IMAP4` constructor.  The connection objects established by this
-   method will be used in the :meth:`IMAP4.read`, :meth:`IMAP4.readline`,
-   :meth:`IMAP4.send`, and :meth:`IMAP4.shutdown` methods.
-   The optional *timeout* parameter specifies a timeout in seconds
-   for the connection attempt. If timeout is not given or is None,
-   the global default socket timeout is used.
-   You may override this method.
+   Opens socket to *port* at *host*. The optional *timeout* parameter
+   specifies a timeout in seconds for the connection attempt.
+   If timeout is not given or is None, the global default socket timeout
+   is used. This method is implicitly called by the :class:`IMAP4` constructor.
+   The connection objects established by this method will be used in
+   the :meth:`IMAP4.read`, :meth:`IMAP4.readline`, :meth:`IMAP4.send`,
+   and :meth:`IMAP4.shutdown` methods. You may override this method.
 
    .. audit-event:: imaplib.open self,host,port imaplib.IMAP4.open
 

--- a/Doc/whatsnew/3.9.rst
+++ b/Doc/whatsnew/3.9.rst
@@ -172,6 +172,9 @@ imaplib
 
 :class:`~imaplib.IMAP4` and :class:`~imaplib.IMAP4_SSL` now have
 an optional *timeout* parameter for their constructors.
+Also, the :meth:`~imaplib.IMAP4.open` method now has an optional *timeout* parameter
+with this change. The overridden methods of :class:`~imaplib.IMAP4_SSL` and
+:class:`~imaplib.IMAP4_stream` were applied to this change.
 (Contributed by Dong-hee Na in :issue:`38615`.)
 
 os

--- a/Lib/imaplib.py
+++ b/Lib/imaplib.py
@@ -140,6 +140,8 @@ class IMAP4:
             host - host's name (default: localhost);
             port - port number (default: standard IMAP4 port).
             timeout - socket timeout (default: None))
+                      If timeout is not given or is None,
+                      the global default socket timeout is used
 
     All IMAP4rev1 commands are supported by methods of the same
     name (in lower-case).
@@ -1274,7 +1276,8 @@ if HAVE_SSL:
                               and private key (default: None)
                 Note: if ssl_context is provided, then parameters keyfile or
                 certfile should not be set otherwise ValueError is raised.
-                timeout - socket timeout (default: None)
+                timeout - socket timeout (default: None) If timeout is not given or is None,
+                          the global default socket timeout is used
 
         for more documentation see the docstring of the parent class IMAP4.
         """

--- a/Lib/imaplib.py
+++ b/Lib/imaplib.py
@@ -291,12 +291,16 @@ class IMAP4:
         # Default value of IMAP4.host is '', but socket.getaddrinfo()
         # (which is used by socket.create_connection()) expects None
         # as a default value for host.
+        if timeout is not None and not timeout:
+            raise ValueError('A non-blocking socket is not supported')
         host = None if not self.host else self.host
         sys.audit("imaplib.open", self, self.host, self.port)
-        timeout = socket._GLOBAL_DEFAULT_TIMEOUT if timeout is None else timeout
-        return socket.create_connection((host, self.port), timeout)
+        address = (host, self.port)
+        if timeout is not None:
+            return socket.create_connection(address, timeout)
+        return socket.create_connection(address)
 
-    def open(self, host = '', port = IMAP4_PORT, timeout = None):
+    def open(self, host='', port=IMAP4_PORT, timeout=None):
         """Setup connection to remote server on "host:port"
             (default: localhost:standard IMAP4 port).
         This connection will be used by the routines:
@@ -1335,7 +1339,7 @@ class IMAP4_stream(IMAP4):
         IMAP4.__init__(self)
 
 
-    def open(self, host = None, port = None):
+    def open(self, host=None, port=None, timeout=None):
         """Setup a stream connection.
         This connection will be used by the routines:
             read, readline, send, shutdown.

--- a/Lib/test/test_imaplib.py
+++ b/Lib/test/test_imaplib.py
@@ -443,15 +443,14 @@ class NewIMAPTestsMixin():
     def test_imaplib_timeout_test(self):
         _, server = self._setup(SimpleIMAPHandler)
         addr = server.server_address[1]
-        client = self.imap_class("localhost", addr, timeout=10.0)
-        self.assertEqual(client.sock.timeout, 10.0)
+        client = self.imap_class("localhost", addr, timeout=support.LOOPBACK_TIMEOUT)
+        self.assertEqual(client.sock.timeout, support.LOOPBACK_TIMEOUT)
         client.shutdown()
         client = self.imap_class("localhost", addr, timeout=None)
         self.assertEqual(client.sock.timeout, None)
         client.shutdown()
-        client = self.imap_class("localhost", addr, timeout=socket._GLOBAL_DEFAULT_TIMEOUT)
-        self.assertEqual(client.sock.timeout, None)
-        client.shutdown()
+        with self.assertRaises(ValueError):
+            client = self.imap_class("localhost", addr, timeout=0)
 
     def test_imaplib_timeout_functionality_test(self):
         class TimeoutHandler(SimpleIMAPHandler):
@@ -462,7 +461,7 @@ class NewIMAPTestsMixin():
         _, server = self._setup(TimeoutHandler)
         addr = server.server_address[1]
         with self.assertRaises(socket.timeout):
-            client = self.imap_class("localhost", addr, timeout=0.1)
+            client = self.imap_class("localhost", addr, timeout=0.001)
 
     def test_with_statement(self):
         _, server = self._setup(SimpleIMAPHandler, connect=False)

--- a/Misc/NEWS.d/next/Library/2019-11-17-17-32-35.bpo-38615.OVyaNX.rst
+++ b/Misc/NEWS.d/next/Library/2019-11-17-17-32-35.bpo-38615.OVyaNX.rst
@@ -1,2 +1,5 @@
 :class:`~imaplib.IMAP4` and :class:`~imaplib.IMAP4_SSL` now have an 
-optional *timeout* parameter for their constructors. Patch by Dong-hee Na.
+optional *timeout* parameter for their constructors. 
+Also, the :meth:`~imaplib.IMAP4.open` method now has an optional *timeout* parameter
+with this change. The overridden methods of :class:`~imaplib.IMAP4_SSL` and
+:class:`~imaplib.IMAP4_stream` were applied to this change. Patch by Dong-hee Na.

